genie-common/src/main/java/com/netflix/genie/common/dto/v4/CommonUserMetadata.java

```diff
@@ -0,0 +1,193 @@
+/*
+ *
+ *  Copyright 2018 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.common.dto.v4;
+
+import com.fasterxml.jackson.annotation.JsonSetter;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.collect.ImmutableSet;
+import com.netflix.genie.common.exceptions.GeniePreconditionException;
+import com.netflix.genie.common.util.GenieObjectMapper;
+import lombok.Data;
+
+import javax.annotation.Nullable;
+import javax.validation.constraints.NotEmpty;
+import javax.validation.constraints.Size;
+import java.io.IOException;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Metadata fields common to all Genie resources (Jobs, clusters, etc).
+ *
+ * @author tgianos
+ * @since 4.0.0
+ */
+@Data
+public abstract class CommonUserMetadata {
+    @NotEmpty(message = "A name is required and must be at most 255 characters")
+    @Size(max = 255, message = "The name can be no longer than 255 characters")
+    private final String name;
+    @NotEmpty(message = "A user is required and must be at most 255 characters")
+    @Size(max = 255, message = "The user can be no longer than 255 characters")
+    private final String user;
+    @Size(max = 255, message = "The version can be no longer than 255 characters")
+    private final String version;
+    @Size(max = 1000, message = "The description can be no longer than 1000 characters")
+    private final String description;
+    private final JsonNode metadata;
+    private final ImmutableSet<@Size(max = 255, message = "A tag can't be longer than 255 characters") String> tags;
+
+    @SuppressWarnings("unchecked")
+    CommonUserMetadata(final Builder builder) {
+        this.name = builder.bName;
+        this.user = builder.bUser;
+        this.version = builder.bVersion;
+        this.description = builder.bDescription;
+        this.metadata = builder.bMetadata;
+        this.tags = builder.bTags == null ? ImmutableSet.of() : ImmutableSet.copyOf(builder.bTags);
+    }
+
+    /**
+     * Get the version.
+     *
+     * @return The version as an {@link Optional}
+     */
+    public Optional<String> getVersion() {
+        return Optional.ofNullable(this.version);
```

Comment:
Given that `Optional` are immutable, why not create them at construction time?
Would create fewer runtime objects and could generate getters via annotation.

Reply:
I was thinking about this too while I was coding it. Generally I've tried to avoid doing what you say because `Optional` best practice articles I've read from community say only really use them to return from methods because if you store them in internal state you have to unwrap them all the time. However in this instance since the classes are entirely immutable and just used to transfer data around I think what you're saying is the better thing to do. I'll make the change for the v4 DTOs in this PR

Reply:
So I went to change this and IntelliJ gave the following warning:
```
Reports any uses of java.util.Optional<T>, java.util.OptionalDouble, java.util.OptionalInt, java.util.OptionalLong or com.google.common.base.Optional as the type for a field or a parameter. Optional was designed to provide a limited mechanism for library method return types where there needed to be a clear way to represent "no result". Using a field with type java.util.Optional is also problematic if the class needs to be Serializable, which java.util.Optional is not. 
```

The first part of that Is what I was commenting on above but we both agree that's not really a big deal in this case. The last sentence though is interesting. What do you think? Force it or just leave it as is and deal with the small overhead?

Reply:
Mh, dunno. Sounds like they have a strong opinion against `Optional` here...